File: 13-RomanToInteger/13-RomanToInteger.py
# Last updated: 12/6/2025, 5:55:22 am
class Solution:
    def romanToInt(self, s: str) -> int:
        ans = 0
        arr = ['I', 'V', 'X', "L", 'C', 'D', 'M']
        val = [1, 5, 10, 50, 100, 500, 1000]
        hash = {
            'I': 0,
            'V': 1,
            'X': 2,
            'L': 3,
            'C': 4,
            'D': 5,
            'M': 6
        }

        spec = {
            'I': 0, 'X': 0, 'C': 0
        }

        n = len(s)
        i = 0
        while i < n:
            index = hash[s[i]]
            if s[i] in spec and i+1 < n:
                if s[i+1] == arr[index+1]:
                    ans += val[index+1] - val[index]
                    i += 1
                elif s[i+1] == arr[index+2]:
                    ans += val[index+2] - val[index]
                    i += 1
                else:
                    ans += val[index]
            else:
                ans += val[index]
            i += 1

        # "I", 'II', 'III', 'IV', 'V', 'VI', 'VII', 'VIII', 'IX', 'X'
        return ans

File: 13-RomanToInteger/13-RomanToInteger.cpp
// Last updated: 12/6/2025, 5:55:27 am
class Solution {
public:
    int checkIf(vector<char> &symb, char c){
        for (int i=0; i<symb.size(); i++){
            if (symb[i] == c) return i;
        }return -1;
    }
    int romanToInt(string s) {
        vector<int> nums = {1,5,10,50,100,500,1000};
        vector<char> symbols = {'I','V','X','L','C','D','M'};
        int ret = 0;
        for (int i=s.length()-1; i>-1; i--){
            int tmp = checkIf(symbols, s[i]);
            if (tmp-1 >= 0 && s[i-1] == symbols[tmp-1]){
                ret += (nums[tmp] - nums[tmp-1]);
                i--;
            }else if (tmp+1 < nums.size() && s[i-1] == symbols[tmp+1]){
                ret += (nums[tmp] + nums[tmp+1]);
                i--;
            }else if (tmp-2 >= 0 && s[i-1] == symbols[tmp-2]){
                ret += (nums[tmp] - nums[tmp-2]);
                i--;
            }else ret += nums[tmp];
        }return ret;
    }
};